### PR TITLE
test: Add ability to trigger verification after review

### PR DESCRIPTION
**Pull request by stefwalter**
We can use the ```test/trigger-verify``` to trigger pull requests that the verifiers don't pick up automatically.

Also move the tool style logic out of testinfra, that's really for shared code.

### CODE REVIEW

**Comment by stefwalter** [comment]
I'd like to make this so it can also trigger new operating systems on a pull request that would otherwise not be implemented. Will remove needswork when that's done.

**Comment by stefwalter** [comment]
Okay, this should be ready now.

**Comment by stefwalter** [comment]
This is blocking testing of #3313

File: test/testinfra.py [hunk 1]
```diff
@@ -193,6 +175,22 @@ def __init__(self, base="/repos/cockpit-project/cockpit/"):
                 raise
         self.available = self.token and True or False
 
+        # Try to load the whitelists
```
Comment (dperpeet) [inline test/testinfra.py]: Why did you move this code out of the function? It's more than a few lines and performs a specific task.

Comment (stefwalter) [inline test/testinfra.py]: Because it's configuration, just like the token above.

File: test/testinfra.py [hunk 2]
```diff
@@ -325,26 +322,32 @@ def update_status(revision, context, last, changes):
...
+                        continue
+                    status = { }
 
-                number = pull["number"]
-                labels = self.labels(number)
-                revision = pull["head"]["sha"]
-                statuses = self.statuses(revision)
+                # For unmarked untested status, user must be in whitelist
+                if login not in self.whitelist:
+                    if status.get("description", NO_TESTING) == NO_TESTING:
```
Comment (dperpeet) [inline test/testinfra.py]: These conditions could be combined with `and`. Also, I think this warrants a comment, e.g. "# If testing has been enabled previously for this, allow further automatic testing"
Note that this is a different level of security, but I'm fine with this. It effectively means whitelisting a user/pull request combination for any future pushes.

File: test/trigger-verify [hunk 3]
```diff
@@ -12,10 +12,35 @@ def main():
...
+    statuses = github.statuses(revision)
+    status = statuses.get(context, { })
+    if status.get("state", "empty") not in ["empty", "error", "failure"]:
+        if opts.force:
+            sys.stderr.write("Pull request isn't in error state, but forcing update.\n")
+        else:
+            sys.stderr.write("Pull request isn't in error state. Status is: '{0}'\n".format(status["state"]))
+            return 1
+
+    changes = { "state": "pending", "description": testinfra.NOT_TESTED, "context": context }
```
Comment (dperpeet) [inline test/trigger-verify]: I wonder if we should add a mode here to disable testing again. Follow-up once it's actually needed?

File: test/testinfra.py (hunk 2 above)
Comment (stefwalter) [inline test/testinfra.py]: That's actually a likely misunderstanding, so I should add a comment here.

It's only for the given commit. Since statuses are per commit. Once the user pushes a new commit into the pull request, it must be triggered for testing again. It's not whitelisting of the login or pull request.

File: test/trigger-verify (hunk 3 above)
Comment (stefwalter) [inline test/trigger-verify]: Sounds good.

File: test/testinfra.py (hunk 1 above)
Comment (dperpeet) [inline test/testinfra.py]: To be fair, the token code has about half the number of lines. But ok, I can live with this. Sometimes it's easier to debug if you don't have to wade through lots of code to see what's actually being done. `read_whitelist` is very concise.

**Comment by stefwalter** [comment]
Fixed up.

File: test/testinfra.py (hunk 2 above)
Comment (dperpeet) [inline test/testinfra.py]: > That's actually a likely misunderstanding, so I should add a comment here.

I guess I walked right into that one. quod erat demonstrandum? =)

File: test/trigger-verify [hunk 4]
```diff
@@ -12,10 +12,35 @@ def main():
...
+        sys.stderr.write("warning: pull request author '{0}' isn't in github-whitelist.\n".format(login))
+
+    revision = pull['head']['sha']
+    statuses = github.statuses(revision)
+    status = statuses.get(context, { })
+    if status.get("state", "empty") not in ["empty", "error", "failure"]:
+        if opts.force:
+            sys.stderr.write("Pull request isn't in error state, but forcing update.\n")
+        else:
+            sys.stderr.write("Pull request isn't in error state. Status is: '{0}'\n".format(status["state"]))
```
Comment (dperpeet) [inline test/trigger-verify]: this will be triggered if state is pending but with description NO_TESTING (Manual testing required)

Comment (dperpeet) [inline test/trigger-verify]: we can change this in a followup, if necessary

for now, we can just use -f
